criu/pages-compress.c

```diff
@@ -0,0 +1,208 @@
+#include <limits.h>
+#include <pthread.h>
+#include <sched.h>
+#include <stdio.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include <compel/plugins/std/syscall-codes.h>
+
+#include "pages-compress.h"
+
+#include "crtools.h"
+#include "image.h"
+#include "log.h"
+#include "memfd.h"
+#include "sched.h"
+#include "util.h"
+
+#include <lz4/programs/lz4io.h>
+#include <lz4/lib/lz4frame.h>
+
+#undef LOG_PREFIX
+#define LOG_PREFIX "decompress: "
+
+const char *temp_path_decompressed = "/tmp/decompressed";
+
+int compress_images(void)
+{
+	char format[PATH_MAX];
+	char pathsrc[PATH_MAX];
+	char pathdst[PATH_MAX];
+	const long pid = getpid();
+	const int dfd = get_service_fd(IMG_FD_OFF);
+	const int id = 1; // the only image we'd like to compress
+	int ret;
+
+	LZ4IO_prefs_t *lz4_prefs = LZ4IO_defaultPreferences();
+	LZ4IO_setBlockSize(lz4_prefs, 128 * 1024);
+
+	pr_debug("Compressing pages\n");
+
+	sprintf(format, "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES].fmt);
+	sprintf(pathsrc, format, pid, dfd, id);
+
+	sprintf(format, "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES_COMP].fmt);
+	sprintf(pathdst, format, pid, dfd, id);
+
+	ret = LZ4IO_compressFilename(pathsrc, pathdst, 8, lz4_prefs);
+	LZ4IO_freePreferences(lz4_prefs);
+
+	if (ret) {
+		pr_err("Can't compress %s to %s\n", pathsrc, pathdst);
+		return -1;
+	}
+	if (unlink(pathsrc)) {
+		pr_perror("Can't delete uncompressed source pages image: %s\n", pathsrc);
+		return -1;
+	}
+	return 0;
+}
+
+static int decompress_image(int comp_fd) {
+
+	const int compbufsize = 64 * 1024;
+	char compbuf[compbufsize];
+	const int outbufsize = 4 * 1024 * 1024; // LZ4F_max4MB is the max block size; see LZ4F_blockSizeID_t for details
+	char outbuf[outbufsize];
+	LZ4F_errorCode_t lz4err;
+	LZ4F_decompressionContext_t dctx;
+	int ret;
+	size_t totalread = 0;
+	size_t totalwrite = 0;
+	int bytestoread = compbufsize;
+	size_t offset = 0;
+
+	lz4err = LZ4F_createDecompressionContext(&dctx, LZ4F_VERSION);
+	if (LZ4F_isError(lz4err)) {
+		pr_err("Can't create LZ4 decompression context\n");
+		return -1;
+	}
+
+	pr_debug("Creating file at '%s'\n", temp_path_decompressed);
+	ret = open(temp_path_decompressed, O_CREAT | O_WRONLY | O_TRUNC, 0600);
+
+	if (0 > ret) {
+		pr_err("Can't create file, errno=%d\n", errno);
+		LZ4F_freeDecompressionContext(dctx);
+		return -1;
+	}
+
+	while (1) {
+		const size_t readbytes = read(comp_fd, compbuf + offset, bytestoread);
+		totalread += readbytes;
+		if (!readbytes && 0 == offset) {
+			/* reached end of file or stream */
+			break;
+		}
+		{
+			size_t outsize = sizeof outbuf;
+			size_t insize = offset + readbytes;
+			const size_t insize_orig = insize;
+			lz4err = LZ4F_decompress(dctx, outbuf, &outsize, compbuf, &insize, NULL);
+			if (LZ4F_isError(lz4err)) {
+				pr_err("LZ4 Decompress error: %s\n", LZ4F_getErrorName(lz4err));
+				break;
+			}
+			{
+				const ssize_t res = write(ret, outbuf, outsize);
```

Comment:
One could also `mmap()` the whole decompressed output file as we know how big it will be and decompress it to the mmapped pages.
But then there is the obvious question which must have been considered - why to write the decompressed pages to file when they will be just `mmap()ed` back into memory? criu has a bit complex mapping functionality but maybe there could be a shortcut for the most common case to just call `LZ4F_decompress()` and decompress the pages to their final address. That can be sure implemented on top of this patch as another acceleration.

Reply:
Thank you, you're right in general. The changes seem to be too complicated if implement all at once, so I decided to postpone using direct memory area for a decompressed file for further iterations.